### PR TITLE
Make chisel a build snap

### PR DESCRIPTION
**Pull request by linostar**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Chisel isn't wanted to be embedded in rockcraft snap, so it is set to be a build-snap instead.

### CODE REVIEW

**Comment by linostar** [comment]
Here are successful examples of running `rockcraft pack` after this PR change:

1. ["Hello World" tutorial](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/tutorials/hello-world/) (rockcraft.yaml that has slices)

```
Starting rockcraft                                                                                                                                                                                        
Logging execution to '/root/.local/state/rockcraft/log/rockcraft-20240126-081459.767870.log'                                                                                                              
Launching managed ubuntu 22.04 instance...                                                                                                                                                                
Creating instance from base instance                                                                                                                                                                      
Starting instance                                                                                                                                                                                         
Starting rockcraft                                                                                                                                                                                        
Logging execution to '/tmp/rockcraft.log'                                                                                                                                                                 
Extracting bare:latest                                                                                                                                                                                    
Extracted bare:latest                                                                                                                                                                                     
Initialising lifecycle                                                                                                                                                                                    
Installing build-packages                                                                                                                                                                                 
Pulling hello                                                                                                                                                                                             
Fetching stage-packages                                                                                                                                                                                   
Pulling pebble                                                                                                                                                                                            
Overlaying hello                                                                                                                                                                                          
Overlaying pebble                                                                                                                                                                                         
Building hello                                                                                                                                                                                            
Building pebble                                                                                                                                                                                           
Staging hello                                                                                                                                                                                             
Staging pebble                                                                                                                                                                                            
Priming hello                                                                                                                                                                                             
Priming pebble                                                                                                                                                                                            
:: + craftctl default                                                                                                                                                                                     
:: + mkdir -p var/lib/pebble/default/layers                                                                                                                                                               
Packing...                                                                                                                                                                                                
Creating new layer                                                                                                                                                                                        
Created new layer                                                                                                                                                                                         
Adding Pebble entrypoint                                                                                                                                                                                  
Configuring entrypoint...                                                                                                                                                                                 
Entrypoint set to ['/bin/pebble', 'enter', '--verbose']                                                                                                                                                   
Adding metadata                                                                                                                                                                                           
Configuring labels and annotations...                                                                                                                                                                     
Labels and annotations set to ['org.opencontainers.image.version=1.0', 'org.opencontainers.image.title=hello', 'org.opencontainers.image.ref.name=hello', 'org.opencontainers.image.licenses=Apache-2.0', 'org.opencontainers.image.created=2024-01-26T08:16:08.731117+00:00', 'org.opencontainers.image.base.digest=08dd2aa57bdccd24cd3342b1a8b4f024bc63ce995411a189dbe9d9e9d0f954e5']                              
Setting the rock's control data                                                                                                                                                                           
Control data written                                                                                                                                                                                      
Metadata added                                                                                                                                                                                            
Exporting to OCI archive                                                                                                                                                                                  
Exported to OCI archive 'hello_1.0_amd64.rock'                                                                                                                                                            
Packed hello_1.0_amd64.rock 
```

2. [Node.js app tutorial](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/tutorials/node-app/) (rockcraft.yaml without slices)

```
Starting rockcraft
Logging execution to '/root/.local/state/rockcraft/log/rockcraft-20240126-082323.896225.log'
Launching managed ubuntu 22.04 instance...
Creating instance from base instance
Starting instance
Starting rockcraft
Logging execution to '/tmp/rockcraft.log'
Retrieving base ubuntu@22.04 for amd64
Retrieved base ubuntu@22.04 for amd64
Extracting ubuntu:22.04
Extracted ubuntu:22.04
Initialising lifecycle
Installing build-packages
Pulling app
Pulling pebble
Overlaying app
Overlaying pebble
Building app
:: + '[' '!' -f /root/parts/app/install/bin/node ']'
:: + curl -s https://nodejs.org/dist/v21.1.0/node-v21.1.0-linux-x64.tar.gz
:: + tar xzf - -C /root/parts/app/install/ --no-same-owner --strip-components=1
:: ++ npm pack .
:: ++ tail -1
:: npm notice
:: npm notice 📦  node_web_app@1.0.0
:: npm notice === Tarball Details ===                                                                                                                                                                     
:: npm notice name:          node_web_app                                                                                                                                                                 
:: npm notice version:       1.0.0                                                                                                                                                                        
:: npm notice filename:      node_web_app-1.0.0.tgz                                                                                                                                                       
:: npm notice package size:  458 B                                                                                                                                                                        
:: npm notice unpacked size: 569 B                                                                                                                                                                        
:: npm notice shasum:        4b08267096e1706a4e4b686641c5d6abc72e3cd9                                                                                                                                     
:: npm notice integrity:     sha512-grSoUXqxxCaAi[...]fILWatiXa/Q9g==                                                                                                                                     
:: npm notice total files:   2                                                                                                                                                                            
:: npm notice                                                                                                                                                                                             
:: + npm install -g --prefix /root/parts/app/install node_web_app-1.0.0.tgz                                                                                                                               
::                                                                                                                                                                                                        
:: added 63 packages in 3s                                                                                                                                                                                
::                                                                                                                                                                                                        
:: 11 packages are looking for funding                                                                                                                                                                    
::   run `npm fund` for details                                                                                                                                                                           
:: npm notice                                                                                                                                                                                             
:: npm notice New minor version of npm available! 10.2.0 -> 10.4.0                                                                                                                                        
:: npm notice Changelog: <https://github.com/npm/cli/releases/tag/v10.4.0>                                                                                                                                
:: npm notice Run `npm install -g npm@10.4.0` to update!                                                                                                                                                  
:: npm notice                                                                                                                                                                                             
Building pebble                                                                                                                                                                                           
Staging app                                                                                                                                                                                               
Staging pebble                                                                                                                                                                                            
Priming app                                                                                                                                                                                               
Priming pebble                                                                                                                                                                                            
:: + craftctl default                                                                                                                                                                                     
:: + mkdir -p var/lib/pebble/default/layers                                                                                                                                                               
Packing...                                                                                                                                                                                                
Creating new layer                                                                                                                                                                                        
Created new layer                                                                                                                                                                                         
Adding Pebble entrypoint                                                                                                                                                                                  
Configuring entrypoint...                                                                                                                                                                                 
Entrypoint set to ['/bin/pebble', 'enter', '--verbose']                                                                                                                                                   
Configuring Pebble services app                                                                                                                                                                           
Found 0 Pebble layers in the base's root filesystem                                                                                                                                                       
Preparing new Pebble layer file 001-my-node-app.yaml                                                                                                                                                      
Writing new Pebble layer file                                                                                                                                                                             
Adding metadata                                                                                                                                                                                           
Configuring labels and annotations...                                                                                                                                                                     
Labels and annotations set to ['org.opencontainers.image.version=1.0', 'org.opencontainers.image.title=my-node-app', 'org.opencontainers.image.ref.name=my-node-app', 'org.opencontainers.image.licenses=GPL-3.0', 'org.opencontainers.image.created=2024-01-26T08:25:04.369867+00:00', 'org.opencontainers.image.base.digest=1ea0c67060c5054fa74789f089d27f01442588e8874159de1c7aa61fb6bf70fc']                     
Setting the rock's control data                                                                                                                                                                           
Control data written                                                                                                                                                                                      
Metadata added                                                                                                                                                                                            
Exporting to OCI archive                                                                                                                                                                                  
Exported to OCI archive 'my-node-app_1.0_amd64.rock'                                                                                                                                                      
Packed my-node-app_1.0_amd64.rock
```

**Comment by linostar** [comment]
@cjdcordeiro @tigarmo This is ready for review. See [examples](https://github.com/canonical/rockcraft/pull/464#issuecomment-1911659283) of running it above.

Before merging this, [PR #637](https://github.com/canonical/craft-parts/pull/637) needs to be merged first, and function `part_has_slices` need to be removed from rockcraft afterwards in the current PR.

File: rockcraft/services/lifecycle.py
Comment (cjdcordeiro) [inline rockcraft/services/lifecycle.py:52]: still within this PR right? Since we're waiting for https://github.com/canonical/craft-parts/pull/637/

File: rockcraft/services/lifecycle.py
Comment (cjdcordeiro) [inline rockcraft/services/lifecycle.py:61]: I guess this will also slightly change after https://github.com/canonical/craft-parts/pull/637/

**Review by cjdcordeiro — APPROVED** [review]
Nice. Short and sweet. I'm adding a label to be explicit that https://github.com/canonical/craft-parts/pull/637/ should be merged first

File: rockcraft/services/lifecycle.py
Comment (linostar) [reply rockcraft/services/lifecycle.py:52]: yes

File: rockcraft/services/lifecycle.py
Comment (linostar) [reply rockcraft/services/lifecycle.py:61]: yes, it will become `craft_parts.part_has_slices` instead.

File: rockcraft/services/lifecycle.py
Comment (tigarmo) [inline rockcraft/services/lifecycle.py:65]: I'm a bit worried about using a "dynamic" `candidate` snap here. Is there a plan/timeline for `stable` chisel?

Comment (linostar) [reply rockcraft/services/lifecycle.py:65]: Not before 13th of Feb, but most likely towards the end of that month.

Comment (linostar) [reply rockcraft/services/lifecycle.py:65]: I had a discussion with Cris about this, and he said it's ok if we wait until chisel stable is released before we merge this one.

File: rockcraft/services/lifecycle.py
Comment (cmatsuoka) [reply rockcraft/services/lifecycle.py:61]: We can move detection and chisel installing to craft-parts, it already does that with other tools such as git.

**Comment by clay-lake** [comment]
Hey everyone. I wanted to check in on this PR; I'm worried its getting a little stale.
Should we keep it open?

:bat:

**Comment by linostar** [comment]
> Hey everyone. I wanted to check in on this PR; I'm worried its getting a little stale. Should we keep it open?
> 
> 🦇

We're still waiting for chisel to release its first stable release, so I guess let's keep it open until then.

File: docs/how-to/code/install-slice/rockcraft.yaml
Comment (lengau) [inline docs/how-to/code/install-slice/rockcraft.yaml:17]: Shouldn't this be `build-snaps`?

Comment (cjdcordeiro) [reply docs/how-to/code/install-slice/rockcraft.yaml:17]: Yep think so

**Comment by cjdcordeiro** [comment]
Chisel is now owned by canonical (@rebornplusplus ) so this PR can be resumed

**Comment by cjdcordeiro** [comment]
We also need this PR (or a new release) for 1.0.0 to become available in rockcraft

**Pull request closed without merging**
